exporter/opentelemetry-exporter-otlp-proto-common/src/opentelemetry/exporter/otlp/proto/common/metrics_encoder/__init__.py

```diff
@@ -0,0 +1,197 @@
+                            min=data_point.min,
+                        )
+                        pb2_metric.histogram.aggregation_temporality = (
+                            metric.data.aggregation_temporality
+                        )
+                        pb2_metric.histogram.data_points.append(pt)
+
+                elif isinstance(metric.data, Sum):
+                    for data_point in metric.data.data_points:
+                        pt = pb2.NumberDataPoint(
+                            attributes=_translate_attributes(
+                                data_point.attributes
+                            ),
+                            start_time_unix_nano=(
+                                data_point.start_time_unix_nano
+                            ),
+                            time_unix_nano=data_point.time_unix_nano,
+                        )
+                        if isinstance(data_point.value, int):
+                            pt.as_int = data_point.value
+                        else:
+                            pt.as_double = data_point.value
+                        # note that because sum is a message type, the
+                        # fields must be set individually rather than
+                        # instantiating a pb2.Sum and setting it once
+                        pb2_metric.sum.aggregation_temporality = (
+                            metric.data.aggregation_temporality
+                        )
+                        pb2_metric.sum.is_monotonic = metric.data.is_monotonic
+                        pb2_metric.sum.data_points.append(pt)
+                else:
+                    _logger.warning(
+                        "unsupported data type %s",
+                        metric.data.__class__.__name__,
+                    )
+                    continue
+
+                pb2_scope_metrics.metrics.append(pb2_metric)
+
+    resource_data = []
+    for (
+        sdk_resource,
+        scope_data,
+    ) in resource_metrics_dict.items():
+        resource_data.append(
+            pb2.ResourceMetrics(
+                resource=Resource(
+                    attributes=_translate_attributes(sdk_resource.attributes)
+                ),
+                scope_metrics=scope_data.values(),
+            )
+        )
+    resource_metrics = resource_data
+    return ExportMetricsServiceRequest(resource_metrics=resource_metrics)
+
+
+def _translate_value(value: Any) -> KeyValue:
+    if isinstance(value, bool):
+        return AnyValue(bool_value=value)
+    elif isinstance(value, str):
+        return AnyValue(string_value=value)
+    elif isinstance(value, int):
+        return AnyValue(int_value=value)
+    elif isinstance(value, float):
+        return AnyValue(double_value=value)
+    elif isinstance(value, Sequence):
+        return AnyValue(
+            array_value=ArrayValue(values=[_translate_value(v) for v in value])
+        )
+    elif isinstance(value, Mapping):
```

Comment:
I have a question here, metrics allows for Mappings but logs and tracing does not, is the correct behavior, I can see in the Protobuf spec that they were added at the same time

Reply:
For reference they were added in the same commit https://github.com/open-telemetry/opentelemetry-proto/commit/e43e1abc40428a6ee98e3bfd79bec1dfa2ed18cd

Reply:
I believe at the time of development for traces and logs, there was no support for mapping. Attributes were either sequence or primitive types. This refactoring can just focus on keeping the current behaviour and things such as this can be addressed separately.

Reply:
+1